src/server/scripts/Northrend/Naxxramas/boss_four_horsemen.cpp

```diff
@@ -82,315 +99,510 @@ enum FourHorsemen
+                                minTime = deathTime;
+                            if (!maxTime || deathTime > maxTime)
+                                maxTime = deathTime;
+                        }
+                        else
+                        {
+                            TC_LOG_WARN("scripts", "FourHorsemenAI: Checking for achievement credit but horseman with id %d is not present", uint32(boss));
+                            return 0u;
+                        }
+                    return (getMSTimeDiff(minTime, maxTime) <= 15 * IN_MILLISECONDS) ? 1u : 0u;
+                }
+                default:
+                    return 0u;
+            }
+        }
 
-        void Reset() override
+        void DoAction(int32 action) override
         {
-            if (!encounterActionReset)
-                DoEncounteraction(NULL, false, true, false);
+            switch (action)
+            {
+                case ACTION_BEGIN_MOVEMENT:
+                    me->GetMotionMaster()->MovePoint(1, _initialPath[0], true);
+                    break;
+                case ACTION_BEGIN_FIGHTING:
+                    me->SetCombatPulseDelay(5);
+                    BeginFighting();
+                    _ourMovementFinished = true;
+                    break;
+            }
+        }
 
-            instance->SetData(DATA_HORSEMEN0 + id, NOT_STARTED);
+        void CheckIsMovementFinished()
+        {
+            for (Horseman boss : horsemen)
+                if (Creature* cBoss = getHorsemanHandle(boss))
+                {
+                    if (cBoss->IsAlive() && !cBoss->AI()->GetData(DATA_MOVEMENT_FINISHED))
+                        return;
+                }
+                else
+                {
+                    TC_LOG_WARN("scripts", "FourHorsemenAI: Checking if movement is finished but horseman with id %d is not present", uint32(boss));
+                    return ResetEncounter();
+                }
 
-            me->SetReactState(REACT_AGGRESSIVE);
-            Initialize();
-            _Reset();
+            for (Horseman boss : horsemen)
+                if (Creature* cBoss = getHorsemanHandle(boss))
+                    cBoss->AI()->DoAction(ACTION_BEGIN_FIGHTING);
         }
 
-        bool DoEncounteraction(Unit* who, bool attack, bool reset, bool checkAllDead)
+        void BeginEncounter()
         {
-            Creature* Thane = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_THANE));
-            Creature* Lady = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_LADY));
-            Creature* Baron = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_BARON));
-            Creature* Sir = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_SIR));
+            if (instance->GetBossState(BOSS_HORSEMEN) == IN_PROGRESS)
+                return;
+            if (!instance->CheckRequiredBosses(BOSS_HORSEMEN))
+                return ResetEncounter();
```

Comment:
please use "return;" when the return type is void otherwise it might look like the called function is returning some kind of value.